satellite/satellitedb/updatechecker.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (C) 2019 Storj Labs, Inc.
+// See LICENSE for copying information.
+
+package satellitedb
+
+import (
+	"sync"
+	"time"
+
+	"storj.io/storj/pkg/pb"
+	"storj.io/storj/pkg/storj"
+	"storj.io/storj/satellite/overlay"
+)
+
+// UpdateChecker is an interface that determines whether updates need to be
+// applied to the underlying overlay cache db.
+type UpdateChecker interface {
+	AddressRequiresUpdate(node *pb.Node) bool
+	GetNodeStats(nodeId storj.NodeID, isUp bool) (stats *overlay.NodeStats)
+	SetNodeStats(nodeId storj.NodeID, isUp bool, stats *overlay.NodeStats)
+}
+
+type addressInfo struct {
+	address   string
+	lastIp    string
+	transport pb.NodeTransport
+}
+
+type uptimeInfo struct {
+	isUp       bool
+	lastUptime time.Time
+	stats      *overlay.NodeStats
+}
+
+type CachedChecker struct {
+	addressLock  sync.RWMutex
+	addressCache map[storj.NodeID]*addressInfo
+
+	uptimeLock  sync.RWMutex
+	uptimeCache map[storj.NodeID]*uptimeInfo
+	isUpDur     time.Duration
+}
+
+func NewCachedChecker(isUpDur time.Duration) *CachedChecker {
+	return &CachedChecker{
+		addressCache: make(map[storj.NodeID]*addressInfo),
+		uptimeCache:  make(map[storj.NodeID]*uptimeInfo),
+		isUpDur:      isUpDur,
+	}
+}
+
+// AddressRequiresUpdate returns true if the address should be updated in the
+// underlying db, and false if not. It also keeps the internal cache up-to-date.
+func (c *CachedChecker) AddressRequiresUpdate(node *pb.Node) bool {
+	// There's nothing we can do with a nil node, or nil address, we're just
+	// gonna say don't update it
+	if node == nil {
+		return false
+	}
+
+	address := node.Address
+	if address == nil {
+		address = &pb.NodeAddress{}
+	}
+
+	c.addressLock.RLock()
+	cached, ok := c.addressCache[node.Id]
+	c.addressLock.RUnlock()
+
+	// If it's not in our cache, add it and say to update
+	if !ok ||
+		address.Address != cached.address ||
+		address.Transport != cached.transport ||
+		node.LastIp != cached.lastIp {
+
+		c.addressLock.Lock()
+		c.addressCache[node.Id] = &addressInfo{
+			address:   address.Address,
+			lastIp:    node.LastIp,
+			transport: address.Transport,
+		}
+		c.addressLock.Unlock()
+		return true
+	}
+
+	return false
+}
+
+// GetNodeStats returns cached NodeStats for the supplied nodeId. If the
+// returned stats are not nil the caller should use them as up-to-date stats. If
+// nil, the called should re-cache NodeStats for this nodeId.
+func (c *CachedChecker) GetNodeStats(nodeId storj.NodeID, isUp bool) (stats *overlay.NodeStats) {
+	c.uptimeLock.RLock()
+	cached, ok := c.uptimeCache[nodeId]
+	c.uptimeLock.RUnlock()
+
+	if !ok ||
+		cached.isUp != isUp ||
+		time.Since(cached.lastUptime) > c.isUpDur {
+
+		return nil
+	}
+
+	return cached.stats
+}
+
+func (c *CachedChecker) SetNodeStats(nodeId storj.NodeID, isUp bool, stats *overlay.NodeStats) {
+	c.uptimeLock.Lock()
+	c.uptimeCache[nodeId] = &uptimeInfo{
+		isUp:       isUp,
+		lastUptime: time.Now(),
```

Comment:
maybe this should be time.Now().UTC()?

Reply:
Can you help me understand why?

Reply:
I think since it's stored in memory there's no functional issue with it currently but it opens up a potential bug in the future if someone hooks this up to persistence and doesn't notice we aren't using UTC. Maybe just use UTC to protect against accidental bugs in the future?

Reply:
oh wait, maybe timezones aren't relevant here, but I thought it was sort of a rule of thumb to generally use `time.Now().UTC()` instead of `time.Now()` for consistency.

Reply:
We already are persisting time.Now() in the overcache db. If you look here https://github.com/storj/storj/blob/bd42ab28d671120e777f479f37aa0af3c4c88bc7/satellite/satellitedb/overlaycache.go#L883 you'll see that what we actually persist is time.Now() in the db, so this keeps the cache using the same times that we have in the db.

Reply:
That won't work right in SQLite. SQLite does preserve the time zone when it converts the value to a string, but then if you try to compare other timestamps against it you won't quite get the right thing (because it's just doing string comparisons). If this isn't causing problems in tests yet, it probably will at some point. You may want to fix both locations.